### PR TITLE
#2641. Check that it is still an error to declare more than one extension type member named `_`

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

**Review by eernstg — COMMENTED** [review]
Here's the tricky concept that I almost forgot about: 'A declaration precludes another declaration' (a setter can preclude a method and a method can preclude a setter).

So I'm scrutinizing these tests to see if some of the expected errors shouldn't be expected after all, because one of the conflicting declarations is precluded.

I think we agree on almost all cases.

However, we do have a few cases where I'd say that a static declaration with basename `_` precludes an instance declaration from a superinterface with basename `_` (one of them being a method and the other a setter).

The specification does not make a distinction between the declaration in the extension type which is the starting point being static or instance (so I'm concluding that it could be a static member declaration, or it could be an instance member declaration). The declaration in a superinterface will have to be an instance member declaration because they are the only ones for which we'll say that 'the declaration _DV_ "has" a member named `_`' or anything else about "has a member".

It is possible that we have lots of old co19 tests about extension types and 'preclude', and they have in practice established the rule that a static declaration can't preclude anything. It is also possible that we just didn't covered this case.

In any case, we may need to adjust the specification if the current behavior says that a static declaration can't preclude anything.

It would be great if you can find those old 'precludes' tests such that we can clarify the situation.

**Comment by sgrekhov** [comment]
@eernstg thank you for review!

> It is possible that we have lots of old co19 tests about extension types and 'preclude', and they have in practice established the rule that a static declaration can't preclude anything. It is also possible that we just didn't covered this case.

No we don't have "preclude" tests that include static members. Currently it's an error in both analyzer and CFE
```dart
class ContainsMethod {
  int foo() => 0;
}

class ContainsGetter {
  int get foo => 0;
}

class ContainsSetter {
  void set foo(int v) {}
}

extension type ET1(ContainsMethod id) implements ContainsMethod {
  static int foo = 1; // CFE: Error: Can't declare a member that conflicts with an inherited one.
                      // Analyzer: Class 'ET1' can't define static member 'foo' and have instance member
                      // 'ContainsMethod.foo' with the  same name. Try renaming the member to a name
                      // that doesn't conflict.  conflicting_static_and_instance
}

extension type ET2(ContainsGetter id) implements ContainsGetter {
  static int foo = 1; // Same as the above
}

extension type ET3(ContainsSetter id) implements ContainsSetter {
  static int foo = 1; // Same as the above
}
```
So, I'm going to:
- create issues for the both tools
- fix these wildcards tests
- add missing extension types tests

Or we can update the spec and I'll do only the last point. Please confirm which plan is better

File: LanguageFeatures/Wildcards/other_declarations_A05_t33.dart [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+extension type ET4(int id) {
+  static void set _(int v) {}
+  static void set _(int v) {}
+//                ^
+// [analyzer] unspecified
+// [cfe] unspecified
+}
+
+extension type ET5(int _) {
+  static int _ = 1;
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Wildcards/other_declarations_A05_t33.dart]: Of course the setter expected here. Thank you for the noticing! Fixed

**Comment by eernstg** [comment]
Thanks for checking! I've recommended that we change the specification such that preclusion is never caused by a static declaration, https://github.com/dart-lang/language/pull/3823. With a little luck it's going to be accepted by the rest of the language team soon.

**Comment by eernstg** [comment]
What's more convenient for you? We could close the threads where a preclusion is caused by a static member and land this (that's basically just "land it now, ignoring some comments"), or we could wait for a decision on dart-lang/language#3822?

**Comment by sgrekhov** [comment]
Let's land it. If something will be changed I'll do appropriate changes in a separate PR

File: LanguageFeatures/Wildcards/other_declarations_A05_t30.dart [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+//           ^
+// [analyzer] unspecified
+  int _() => 6;
+//    ^
+// [cfe] unspecified
+}
+
+extension type ET6Implements(ContainsWildcardMethod id)
+      implements ContainsWildcardMethod {
+  static int _ = 1;
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Wildcards/other_declarations_A05_t30.dart:80]: Closing this thread: We assume that static members cannot cause preclusion.

File: LanguageFeatures/Wildcards/other_declarations_A05_t32.dart [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+// [analyzer] unspecified
+// [cfe] unspecified
+  void set _(int v) {}
+//         ^
+// [cfe] unspecified
+}
+
+extension type ET8Implements(ContainsWildcardSetter id)
+      implements ContainsWildcardSetter {
+  static int _() => 0;
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Wildcards/other_declarations_A05_t32.dart:115]: Closing this thread: We assume that static members cannot cause preclusion.

File: LanguageFeatures/Wildcards/other_declarations_A05_t33.dart [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+// [analyzer] unspecified
+// [cfe] unspecified
+  int _() => 6;
+//    ^
+// [cfe] unspecified
+}
+
+extension type ET6Implements(ContainsWildcardMethod id)
+      implements ContainsWildcardMethod {
+  static void set _(int v) {}
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Wildcards/other_declarations_A05_t33.dart:80]: Closing this thread: We assume that static members cannot cause preclusion.

**Review by eernstg — APPROVED** [review]
LGTM!

**Pull request merged**